src/extract-typedefs.ts

```diff
@@ -0,0 +1,179 @@
+  const names = name.split(',').map(s => s.trim()).filter(s => !!s);
+  const children = parseDeclarations(body);
+  const typedef: Partial<CDeclaration> = { children };
+  if (matchedToken === 'union') typedef.type = 'union';
+  names.forEach(name => {
+    typedefs[name] = { name, type: matchedToken, ...typedef };
+  });
+}
+
+function parseDeclarations(body: string): CDeclaration[] {
+  const declarations: CDeclaration[] = [];
+  const lines = body.split('\n');
+  let offset = 0;
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i];
+    const matchRecursiveDecl = line.match(/^(struct|union) /);
+    if (matchRecursiveDecl) {
+      const { body: nextBody, name } = parseNamedBlock(body, offset, matchRecursiveDecl[0].length);
+      const children = parseDeclarations(nextBody);
+      declarations.push({
+        name,
+        children,
+        count: 1,
+        type: matchRecursiveDecl[1],
+      });
+      i += nextBody.split('\n').length - 2;
+    } else {
+      const matchDecl = line.match(DECLARATION_REGEX);
+      if (matchDecl) {
+        const [_, type, name, rawLength, rawComment] = matchDecl;
+        const comment = rawComment ? rawComment.replace(/(^\/\/|^\/\*|\*\/$)/g, '').replace(/\s+/g, ' ').trim() : null;
+        const lengths = rawLength ? rawLength.split(/\D+/).filter(f => !!f).map(i => parseInt(i, 10)) : [1];
+        let stringLength = null;
+        let count: number | number[] = lengths;
+        if (type === 'char') stringLength = lengths.pop();
+        if (Array.isArray(count) && count.length === 1) count = count[0];
+        if (Array.isArray(count) && count.length === 0) count = 1;
+        declarations.push({
+          type,
+          name,
+          count,
+          comment,
+          stringLength,
+        });
+      } else if (declarations.length > 0) {
+        const matchComment = line.match(COMMENT_REGEX);
+        if (matchComment) {
+          declarations[declarations.length - 1].comment += matchComment[1].replace(/\s+/g, ' ').trim();
+        }
+      }
+    }
+    offset += line.length + 1;
+  }
+  return declarations;
+}
+
+function getSize(decl: CDeclaration, typedefs: CDeclarationIndex): number {
+  if (decl.byteLength) return decl.byteLength;
+  if (decl.type === 'char') return decl.stringLength;
+  if (decl.type === 'union') return Math.max(...decl.children.map(c => getSize(c, typedefs)));
+  if (decl.children) return decl.children.reduce((acc: number, c) => acc + (getSize(c, typedefs) * reduceCount(c.count)), 0);
+  if (typedefs[decl.type]) return getSize(typedefs[decl.type], typedefs);
+  return 0;
+}
+
+function applySize(decl: CDeclaration, typedefs: CDeclarationIndex): CDeclaration {
+  const next = {
+    ...decl,
+    byteLength: getSize(decl, typedefs) * reduceCount(decl.count)
+  };
+  if (decl.children) {
+    next.children = decl.children.map(c => applySize(c, typedefs));
+  }
+  return next;
+}
+
+function parseHeaderFile(path: string) {
```

Comment:
Is this home grown? Just wondering if anyone else has solved the problem of generating TS definitions for FFI

Reply:
This is indeed home-grown. I went down quite a rabbit hole with this, trying to pull in various parts of C toolchains in order to parse the header file. As insane as this script is, it seemed like the most reasonable approach after quite a bit of research. Existing tools for analyzing C source code work at low enough of a level that they don't offer a much better experience than just string parsing as is done here.

It's potentially brittle but this is a pretty specific case, so the simplicity of doing string parsing versus traversing an AST seemed worth it.

Reply:
Yeah it's some scary parsing going on, I guess if it works, it works! It's only generating TS definitions so it's not _too_ critical if it broke down the line.